### PR TITLE
Refactor auth flow 

### PR DESCRIPTION
**Pull request by trueleo**
Part of #250  

### Description
 - Authentication and Authorization are merged into single middleware.
 - Added AuthMap for better permission lookup
 - Permissions are loaded lazily into AuthMap once user is authenticated

### CODE REVIEW

**Comment by trueleo** [comment]
I have read the CLA Document and I hereby sign the CLA

File: server/src/handlers/http/middleware.rs [hunk 1]
```diff
@@ -75,25 +76,33 @@ where
...
+    fn call(&self, mut req: ServiceRequest) -> Self::Future {
+        let creds = req.extract::<BasicAuth>().into_inner();
+        let creds: Result<(String, String), Error> = creds.map_err(Into::into).and_then(|creds| {
+            let username = creds.user_id().trim().to_owned();
+            if let Some(password) = creds.password() {
+                Ok((username, password.trim().to_owned()))
+            } else {
+                Err(ErrorBadRequest("Password is required in basic auth"))
+            }
+        });
```
Comment (nitisht) [inline server/src/handlers/http/middleware.rs:88]: need comments in this part - what and why are we doing this

File: server/src/handlers/http/middleware.rs [hunk 2]
```diff
@@ -75,25 +76,33 @@ where
...
-        let extensions = req.extensions();
-        let username = extensions
-            .get::<String>()
-            .expect("authentication layer verified username");
-        let is_auth = Users.check_permission(username, self.action, stream);
-        drop(extensions);
+        let is_auth = creds.map(|creds| {
+            let (username, password) = creds;
+            Users.authenticate(username, password, self.action, stream)
+        });
```
Comment (nitisht) [inline server/src/handlers/http/middleware.rs:97]: what is `is_auth`?

File: server/src/rbac.rs [hunk 3]
```diff
@@ -16,136 +16,139 @@
...
 
+pub mod auth;
 pub mod role;
 pub mod user;
 
-pub static USER_AUTHENTICATION_MAP: OnceCell<RwLock<UserMap>> = OnceCell::new();
-pub static USER_AUTHORIZATION_MAP: OnceCell<RwLock<UserPermMap>> = OnceCell::new();
+pub static USER_MAP: OnceCell<RwLock<UserMap>> = OnceCell::new();
+pub static AUTH_MAP: OnceCell<RwLock<AuthMap>> = OnceCell::new();
+
```
Comment (nitisht) [inline server/src/rbac.rs:37]: What is the purpose of two different maps here?

File: server/src/rbac/auth.rs [hunk 4]
```diff
@@ -0,0 +1,49 @@
+use std::collections::HashMap;
```
Comment (nitisht) [inline server/src/rbac/auth.rs:1]: License

File: server/src/handlers/http/middleware.rs (hunk 2 above)
Comment (trueleo) [reply server/src/handlers/http/middleware.rs:98]: The type is result of authentication and authorization. Type of creds is Result<> which i can only return in the async block down below. So have to `.map` for calling authenticate

`Users.authenticate` does both authentication and authorization.

File: server/src/rbac.rs (hunk 3 above)
Comment (trueleo) [reply server/src/rbac.rs:37]: If any changes made to user ( password change, role update, etc.. ) it is updated in memory in the Usermap `Hashmap<username, User>`. The User contains password hash and role information. While `AuthMap` is for quick authentication and authorization, it is HashMap<(user, pass), Vec<Permission>>. Any changes made to User will reset the AuthMap and upon authentication user perms is reloaded into this map

**Review by nitisht — APPROVED** [review]
LGTM & Tested

**Pull request merged**